### PR TITLE
Factor out langutil, a new static library that is shared between libyul/libsolidity

### PR DESCRIPTION
**Pull request by christianparpart**
Factor out <langutil>, a new [static] library that is shared between libsolidity (solidity language frontend) and future libyul (YUL language frontend)

The motivation behind this is to eliminate circular dependencies in our build system and end up also having a standalone libyul static library (PR following this one).

**NOTE:** This is a recreation of PR #5381 that performs the changes incrementally:
1.  Isolating the files into its own directory (`liblangutil/`)
2. Splitting out new files (for `CharStream` API)
3. Introducing namespace `langutil`
4. Making the new directory its own modular static library.

This PR is also a prerequisite for the #5399 (libyul as static library, eliminating circular dependencies)

## CHANGES (AND THOUGHTS)

* moved general purpose language frontend files out of libsolidity/
* **also** moved Scanner and Token, as they're both shared between Solidity and Yul (for now), can be generalized (maybe later, not now)
* introduced new namespace `langutil` and put all underneath it, adapting the users of them
* also added `EVMVersion` out (as it is shared); however, that one should have a namespace such as `eth`  or `evm` as it's actually about the EVM here (could be more appropriate in another folder, but I didn't wanna overdo it here).
* Extracted CharSource into its own header/impl file, with the future vision of potentially having one lexer/token API per language (solidity, yul,) that could share this API in order allow transition from one recognizer to another (forth and back).

### CODE REVIEW

File: liblangutil/CharStream.h [hunk 1]
```diff
@@ -44,24 +44,34 @@
...
 #include <tuple>
 
-namespace dev
-{
-namespace solidity
-{
+namespace langutil {
 
+/**
+ * Unidirectional stream of characters.
```
Comment (chriseth) [inline liblangutil/CharStream.h]: It actually is bidirectional.

File: liblangutil/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Solidity Commons Library (Solidity related sharing bits between libsolidity and libyul)
+file(GLOB sources "*.cpp")
```
Comment (chriseth) [inline liblangutil/CMakeLists.txt:2]: I think we should start listing files explicitly. It makes it much harder to forget checking in a new file and also avoids problems with .cpp files that are accidentally kept in a directory.

Comment (christianparpart) [reply liblangutil/CMakeLists.txt:2]: Let me do that in a separate PR, please, so I can do that globally on all CMakeList.txt's :)

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 3]
```diff
@@ -36,9 +36,12 @@
 #include <boost/range/adaptor/transformed.hpp>
 
 using namespace std;
-using namespace dev;
-using namespace dev::solidity;
+using namespace langutil;
 
+namespace dev
```
Comment (chriseth) [inline libsolidity/analysis/ReferencesResolver.cpp:41]: Usually we do not have namespaces in .cpp files, but I guess this is to avoid a name clash, right?

**Review by chriseth — DISMISSED** [review]

File: liblangutil/CharStream.h [hunk 4]
```diff
@@ -44,24 +44,34 @@
...
 */
+/**
+ * @author Christian <c@ethdev.com>
+ * @date 2014
+ * Solidity scanner.
+ */
 
 #pragma once
 
+#include <cstdint>
```
Comment (axic) [inline liblangutil/CharStream.h:55]: Is this part of the previous commit?

File: liblangutil/Exceptions.h [hunk 5]
```diff
@@ -98,35 +98,4 @@ class Error: virtual public Exception
 	std::string m_typeName;
 };
 
-
-using errorSourceLocationInfo = std::pair<std::string, SourceLocation>;
-
-class SecondarySourceLocation
```
Comment (axic) [inline liblangutil/Exceptions.h]: Why is this removed?

File: liblangutil/Scanner.h [hunk 6]
```diff
@@ -71,13 +69,12 @@ class Scanner
...
 public:
-
-	explicit Scanner(CharStream const& _source = CharStream(), std::string const& _sourceName = "") { reset(_source, _sourceName); }
+	explicit Scanner(CharStream _source = CharStream(), std::string _sourceName = "") { reset(std::move(_source), std::move(_sourceName)); }
 
 	std::string source() const { return m_source.source(); }
 
 	/// Resets the scanner as if newly constructed with _source and _sourceName as input.
-	void reset(CharStream const& _source, std::string const& _sourceName);
+	void reset(CharStream _source, std::string _sourceName);
```
Comment (axic) [inline liblangutil/Scanner.h:77]: Why aren't these references anymore?

File: liblangutil/Scanner.h [hunk 7]
```diff
@@ -71,13 +69,12 @@ class Scanner
 {
 	friend class LiteralScope;
 public:
-
-	explicit Scanner(CharStream const& _source = CharStream(), std::string const& _sourceName = "") { reset(_source, _sourceName); }
+	explicit Scanner(CharStream _source = CharStream(), std::string _sourceName = "") { reset(std::move(_source), std::move(_sourceName)); }
```
Comment (axic) [inline liblangutil/Scanner.h:72]: Same here.

File: liblangutil/Scanner.h (hunk 6 above)
Comment (chriseth) [reply liblangutil/Scanner.h:77]: Previously, they were copied anyway (inside the function). Now they are copied (or moved) and moved. This is a good style: If you intend to take over the object anyway, pass by value and then move.

In this style, you either have one copy and one move (if called with a non-movable object) or two moves (if called with a movable object).

In the old style, you always have one copy.

**Comment by axic** [comment]
Can I rebase this to move some changes between two commits?

**Comment by chriseth** [comment]
@axic please go ahead! The faster this gets merged, the better.

File: liblangutil/Exceptions.h (hunk 5 above)
Comment (christianparpart) [reply liblangutil/Exceptions.h]: Moved, not removed.

File: liblangutil/Scanner.h (hunk 6 above)
Comment (christianparpart) [reply liblangutil/Scanner.h:77]: It is also the recommended way of passing args in constructors. And you have zero copies for unnamed values (just one move)

Comment (axic) [reply liblangutil/Scanner.h:77]: It is fine, just didn't expect this change to be hidden in this commit :)

File: liblangutil/Exceptions.h (hunk 5 above)
Comment (axic) [reply liblangutil/Exceptions.h]: Actually while digging deeper I don't see the point why this is moved. Also it seems that this class is very poorly designed.

File: libevmasm/ControlFlowGraph.cpp [hunk 8]
```diff
@@ -275,7 +275,7 @@ void ControlFlowGraph::gatherKnowledge()
 			//@todo in the case of JUMPI, add knowledge about the condition to the state
 			// (for both values of the condition)
 			set<u256> tags = state->tagsInExpression(
-				state->stackElement(state->stackHeight(), SourceLocation())
```
Comment (axic) [inline libevmasm/ControlFlowGraph.cpp:278]: These are the other hidden changes in this PR.

File: liblangutil/CharStream.h [hunk 9]
```diff
@@ -56,10 +56,7 @@
 #include <string>
 #include <tuple>
 
-namespace dev
-{
-namespace solidity
-{
+namespace langutil {
```
Comment (axic) [inline liblangutil/CharStream.h]: Style change.

File: liblangutil/Exceptions.h [hunk 10]
```diff
@@ -24,33 +24,33 @@
 
 #include <string>
 #include <utility>
+#include <vector>
+#include <memory>
```
Comment (axic) [inline liblangutil/Exceptions.h:28]: What includes were hiding these? I see a lot of new system headers introduced in this commit.

File: liblangutil/Scanner.cpp [hunk 11]
```diff
@@ -143,10 +141,10 @@ class LiteralScope
 }; // end of LiteralScope class
 
 
-void Scanner::reset(CharStream const& _source, string const& _sourceName)
+void Scanner::reset(CharStream _source, string _sourceName)
 {
-	m_source = _source;
-	m_sourceName = make_shared<string const>(_sourceName);
+	m_source = std::move(_source);
+	m_sourceName = make_shared<string const>(std::move(_sourceName));
```
Comment (axic) [inline liblangutil/Scanner.cpp:147]: I think this changeset would have deserved its own commit as it has nothing to do with namespace changes.

**Comment by axic** [comment]
Fixed a couple of things, but not everything.

In general my issue was that the last commit, titled as introducing `langutil` namespace suggest it is a tiny change, however it also does the following:
- Changes constructor style from `()` to `{}` for sourcelocation (newer C++)
- Introduces a lot of new system includes without explanation
- Changes the API of the Scanner to take source as value (with move) as opposed to as a reference

I think in general it is a bad idea hiding so many unrelated changes in a single commit.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Extended the commit log for the last commit to explain all those other changes. Next time however I think it makes sense to split them properly.

File: liblangutil/Exceptions.h (hunk 10 above)
Comment (christianparpart) [reply liblangutil/Exceptions.h:28]: yes, I wasn't mentioning this in the commit log. sorry. but the if a unit is using a symbol, it should also include the header that is meant to introduce this symbol (I believe I read about that rule in the CppCoreGuidelines, can look for the reference if desired). That's why I was adding them along the way (should be seperate commit/PR).

File: liblangutil/Scanner.cpp (hunk 11 above)
Comment (christianparpart) [reply liblangutil/Scanner.cpp:147]: i agree. (next time) :-)

**Review by chriseth — APPROVED** [review]
